core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1885,7 +1909,27 @@ private int acquireSubsetBatchRecords(
                     continue;
                 }
 
-                InFlightState updateResult =  offsetState.getValue().tryUpdateState(RecordState.ACQUIRED, DeliveryCountOps.INCREASE,
+                int recordDeliveryCount = offsetState.getValue().deliveryCount();
+                // On last delivery attempt, submit acquired records,
+                // bad record will be delivered alone next time
+                if (maxDeliveryCount > 2 && recordDeliveryCount == maxDeliveryCount - 1 && acquiredCount > 0) {
+                    hasBadRecord = true;
+                    break;
+                }
+
+                // When record delivery count reach the throttle threshold, progressively reduce batch size to isolate bad records.
```

Comment:
I won't prefer to use `bad` as we should just talk about the records which have approached the delivery limit and we send individual records.
```suggestion
                // When record delivery count reach the throttle threshold, progressively reduce batch size to isolate records.
```